prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,184 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
+	if (wre->ep.ep) {
+		if ((wre->ep.ep != ep) &&
+		    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
+			wce = util_buf_alloc(cq->wce_pool);
+			if (!wce) {
+				dlist_insert_tail(&wre->entry,
+						  out_list);
+				return -FI_EAGAIN;
+			}
+			memset(wce, 0, sizeof(*wce));
+			wc.wr_id = (uintptr_t)wre->context;
+			wce->wc = wc;
+			slist_insert_tail(&wce->entry, &cq->wcq);
+		}
+	} else {
+		/* if `ep == NULL` it means that WRE is already released */
+		return ret;
+	}
+	tmp_ep = wre->ep.ep;
+	wre->ep.ep = NULL;
+	util_buf_release(tmp_ep->wre_pool, wre);
+	return ret;
+}
+
+void fi_ibv_cleanup_cq(struct fi_ibv_msg_ep *ep)
+{
+	DEFINE_LIST(wre_out_list);
+	int ret;
+	struct fi_ibv_wre *wre;
+	struct dlist_entry *tmp;
+
+	fastlock_acquire(&ep->rcq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->rcq,
+						 &wre_out_list);
+	} while (ret > 0 || ret == -FI_EAGAIN);
+	/* Handle WRs for which there were no appropriate WCs */
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep.ep) &&
+		    (wre->wr.ibv_wr_type == IBV_RECV_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep.ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
+		}
+	}
+	fastlock_release(&ep->rcq->lock);
+
+	fastlock_acquire(&ep->scq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->scq,
+						 &wre_out_list);
+	} while (ret > 0 || ret == -FI_EAGAIN);
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep.ep) &&
+		    (wre->wr.ibv_wr_type == IBV_SEND_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep.ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
+		}
+	}
+	fastlock_release(&ep->scq->lock);
+
+	util_buf_pool_destroy(ep->wre_pool);
+}
+
+/* Must call with cq->lock held */
+ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
+{
+	struct fi_ibv_wre *wre;
+	struct util_buf_pool *wre_pool;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, wc);
+	if (ret <= 0)
+		return ret;
+
+	/* TODO Handle the case when app posts a send with same wr_id */
+	if ((wc->opcode == IBV_WC_RECV) ||
+	    (wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id))
+		goto done;
+
+	return fi_ibv_handle_internal_signal_wc(cq, wc);
+
+done:
+	wre = (struct fi_ibv_wre *)(uintptr_t)wc->wr_id;
+	if (!wre->ep.ep && !wre->ep.srq) {
+		/* it means that WRE is already released */
+		return 0;
+	} else if (wre->ep.ep) {
+		wre_pool = wre->ep.ep->wre_pool;
+		wre->ep.ep = NULL;
+		if (wc->status == IBV_WC_WR_FLUSH_ERR)
+			/* Handles case where remote side destroys
+			 * the connection, but local side isn't aware
+			 * about that yet */
+			ret = 0;
+	} else if (wre->ep.srq) {
+		wre_pool = wre->ep.srq->wre_pool;
+		wre->ep.srq = NULL;
+		if (wc->status == IBV_WC_WR_FLUSH_ERR)
+			/* Handles case where remote side destroys
+			 * the connection, but local side isn't aware
+			 * about that yet */
+			ret = 0;
```

Comment:
Shouldn't we be using ep wre_pool for srq as well? The srq close function assumes that. Here and other places in the code assume separate srq wre pool.

We should just use the EP wre pool everywhere.